runtime/common/QuditIdTracker.h

```diff
@@ -0,0 +1,65 @@
+/*************************************************************** -*- C++ -*- ***
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ *******************************************************************************/
+
+#pragma once
+
+#include <algorithm>
+#include <vector>
+
+namespace cudaq {
+
+/// @brief The QuditIdTracker tracks unique integer
+/// indices for allocated qudits. It will recycle indices
+/// when the qudit is deallocated, so that it can be reused.
+class QuditIdTracker {
+private:
+  /// @brief The current identifier.
+  std::size_t currentId = 0;
+
+  /// @brief The queue of recycled qubits.
+  std::vector<std::size_t> recycledQudits;
+
+public:
+  /// @brief Return the next available index,
```

Comment:
Might want something like:
```cpp
QuditIdTracker() = default;
QutidIdTracker(const QuditIdTracker&) = deleted;
...
```
to prevent accidentally creating copies of this object.